torch_glow/src/GlowIValue.cpp

```diff
@@ -0,0 +1,281 @@
+#include "glow/Support/Error.h"
+#include "glow/Support/Support.h"
+
+namespace glow {
+
+namespace {
+/// \returns a Glow tensor with the same type and underlying data as the given
+/// PyTorch tensor \p ptT.
+// TODO: don't redefine this
+llvm::Expected<glow::Tensor> ptTensorToGlowTensor2(const at::Tensor &ptT) {
+  // TODO: get correct ElemKind.
+  DCHECK_EQ(ptT.scalar_type(), at::kFloat)
+      << "Only float tensors supported currently.";
+  std::vector<size_t> dims;
+  for (const auto d : ptT.sizes()) {
+    dims.push_back(d);
+  }
+
+  glow::Type ty(glow::ElemKind::FloatTy, dims);
+  glow::Tensor glowT(ptT.data_ptr(), &ty);
+
+  return glowT;
+}
+
+const char *tagToStr(GlowIValue::Tag tag) {
+  switch (tag) {
+  case GlowIValue::Tag::None:
+    return "None";
+  case GlowIValue::Tag::Tensor:
+    return "Tensor";
+  case GlowIValue::Tag::Double:
+    return "Double";
+  case GlowIValue::Tag::Int:
+    return "Int";
+  case GlowIValue::Tag::Bool:
+    return "Bool";
+  case GlowIValue::Tag::IntList:
+    return "IntList";
+  case GlowIValue::Tag::DoubleList:
+    return "DoubleList";
+  case GlowIValue::Tag::BoolList:
+    return "BoolList";
+  case GlowIValue::Tag::TensorList:
+    return "TensorList";
+  case GlowIValue::Tag::Tuple:
+    return "Tuple";
+  }
+}
+} // namespace
+
+void GlowIValue::reset() {
+  switch (tag_) {
+  case Tag::Tensor:
+    delete payload_.asTensor;
+    break;
+  case Tag::IntList:
+    delete payload_.asIntList;
+    break;
+  case Tag::DoubleList:
+    delete payload_.asDoubleList;
+    break;
+  case Tag::BoolList:
+    delete payload_.asBoolList;
+    break;
+  case Tag::TensorList:
+    delete payload_.asTensorList;
+    break;
+  case Tag::Tuple:
+    delete payload_.asTuple;
+    break;
+  case Tag::None:
+  case Tag::Double:
+  case Tag::Int:
+  case Tag::Bool:
+    // Nothing to free.
+    break;
+  }
+  tag_ = Tag::None;
+}
+
+GlowIValue::~GlowIValue() { reset(); }
+
+GlowIValue::GlowIValue(GlowIValue &&other) {
+  std::swap(tag_, other.tag_);
+  std::swap(payload_, other.payload_);
+}
+
+GlowIValue &GlowIValue::operator=(GlowIValue &&other) {
+  reset();
+  std::swap(tag_, other.tag_);
+  std::swap(payload_, other.payload_);
+  return *this;
+}
+
+GlowIValue::Tag GlowIValue::getTag() const { return tag_; }
+
+bool GlowIValue::isNone() const { return Tag::None == tag_; }
+bool GlowIValue::isTensor() const { return Tag::Tensor == tag_; }
+bool GlowIValue::isDouble() const { return Tag::Double == tag_; }
+bool GlowIValue::isInt() const { return Tag::Int == tag_; }
+bool GlowIValue::isBool() const { return Tag::Bool == tag_; }
+bool GlowIValue::isIntList() const { return Tag::IntList == tag_; }
+bool GlowIValue::isDoubleList() const { return Tag::DoubleList == tag_; }
+bool GlowIValue::isBoolList() const { return Tag::BoolList == tag_; }
+bool GlowIValue::isTensorList() const { return Tag::TensorList == tag_; }
+
+#define ExpectTag(EXPECTED_TAG)                                                \
```

Comment:
Hard to debug macro, can be just inline function.

Reply:
but this macro inserts a return